### PR TITLE
Feat/transfer reverse

### PR DESCRIPTION
**Pull request by simeonpp**

### CODE REVIEW

File: errors.js [hunk 1]
```diff
@@ -17,6 +17,9 @@ const InvalidPhone = create('invalidPhone', Merchant, 'Invalid phone');
 const InvalidInvoice = create('invalidInvoice', Merchant, 'Invalid invoice');
 const MerchantInsufficientFunds = create('insufficientFunds', Merchant, 'Balance not enough');
 const MerchantUnknown = create('unknown', Merchant, 'Unknown error');
+// reversal
+const Reversal = create('reversal');
```
Comment (kalinkrustev) [inline errors.js]: use NotFound , that is defined above

File: api/sql/schema/750$transfer.push.confirmReversal.sql [hunk 2]
```diff
@@ -23,6 +23,14 @@ BEGIN TRY
     WHERE
         secondTransferId=@transferId
 
+    EXEC [transfer].[push.event]
+        @transferId = @transferId,
+        @type = 'transfer.reverse',
+        @state = 'reversed',
```
Comment (kalinkrustev) [inline api/sql/schema/750$transfer.push.confirmReversal.sql]: use 'reverse' as in other places we use words like 'confirm', 'fail'
also have a look at this fix: https://github.com/softwaregroup-bg/ut-transfer/commit/ac78f86c0438816f850047126549a960697cb16c

File: api/script/index.js [hunk 3]
```diff
@@ -230,14 +254,39 @@ module.exports = {
...
-            .then(result => {
-                if (!result || !result[0] || !result[0][0]) {
-                    throw errors.notFound();
-                } else {
-                    return result;
-                }
-            })
+        var bus = this.bus;
+        var getTransfer = (params) => {
+            if (params.transferId) {
```
Comment (kalinkrustev) [inline api/script/index.js]: we need to discuss this approach, let's do it tomorrow

File: errors.js (hunk 1 above)
Comment (simeonpp) [inline errors.js]: Done.

File: api/sql/schema/750$transfer.push.confirmReversal.sql (hunk 2 above)
Comment (simeonpp) [inline api/sql/schema/750$transfer.push.confirmReversal.sql]: Done.

File: api/script/index.js (hunk 3 above)
Comment (simeonpp) [inline api/script/index.js]: Okey, will discuss it tomorrow.

File: api/sql/schema/750$transfer.transfer.get.sql [hunk 4]
```diff
@@ -17,21 +20,27 @@ AS
    
    SELECT 'transfer' AS resultSetName
    
-   SELECT transferId, transferTypeId, acquirerCode, transferIdAcquirer, transferIdLedger, transferIdIssuer, transferIdMerchant, 
-   transferDateTime, localDateTime, t.settlementDate, channelId, channelType, ordererId, merchantId, merchantInvoice, 
-   merchantPort, merchantType, cardId, sourceAccount, destinationAccount, expireTime, expireCount, reversed, retryTime, 
-   retryCount, ledgerTxState, issuerTxState, acquirerTxState, merchantTxState, issuerId, ledgerId, transferCurrency, transferAmount, 
-   acquirerFee, issuerFee, transferFee, [description], pi.port issuerPort, pl.port ledgerPort
+   SELECT t.transferId, t.transferTypeId, t.acquirerCode, t.transferIdAcquirer, t.transferIdLedger, t.transferIdIssuer, t.transferIdMerchant,
```
Comment (kalinkrustev) [inline api/sql/schema/750$transfer.transfer.get.sql]: SELECT TOP 1 ... ORDER BY t.transferDateTime DESC

File: api/sql/schema/750$transfer.transfer.get.sql [hunk 5]
```diff
@@ -17,21 +20,27 @@ AS
...
    SELECT 'transferSplit' AS resultSetName
 
-   SELECT splitId, ts.transferId, conditionId, splitNameId, debit, credit, amount, ts.[description], tag, debitActorId, creditActorId, debitItemId, creditItemId, [state], [transferIdPayment]
+   SELECT ts.splitId, ts.transferId, ts.conditionId, ts.splitNameId, ts.debit, ts.credit, ts.amount, 
+          ts.[description], ts.tag, ts.debitActorId, ts.creditActorId, ts.debitItemId, ts.creditItemId, 
+          ts.[state], ts.[transferIdPayment]
    FROM [transfer].[split] ts 
    JOIN [transfer].[transfer] tt ON ts.transferId = tt.transferId
-   WHERE (@transferIdAcquirer IS NULL OR [transferIdAcquirer] = @transferIdAcquirer)
+   WHERE (@transferIdAcquirer IS NULL OR tt.[transferIdAcquirer] = @transferIdAcquirer)
```
Comment (kalinkrustev) [inline api/sql/schema/750$transfer.transfer.get.sql]: remove the JOIN and use ts.transferId = transferId acquired from the SELECT TOP 1 query

File: api/script/index.js [hunk 6]
```diff
@@ -9,31 +9,55 @@ var currency = require('../../currency');
...
+            .then(result => {
+                if (reversal.issuerPort === reversal.ledgerPort) {
+                    return reversal;
+                } else {
+                    return portReversal(reversal.ledgerPort, reversal)
+                    .then(() => reversal);
+                }
+            });
+        } else {
+            throw errors.notFound();
```
Comment (kalinkrustev) [inline api/script/index.js]: When transaction was found and was already reversed, we need to throw different error.

File: api/sql/schema/750$transfer.transfer.get.sql (hunk 4 above)
Comment (simeonpp) [inline api/sql/schema/750$transfer.transfer.get.sql]: Done

File: api/sql/schema/750$transfer.transfer.get.sql (hunk 5 above)
Comment (simeonpp) [inline api/sql/schema/750$transfer.transfer.get.sql]: Done.

File: api/sql/schema/750$transfer.transfer.get.sql [hunk 7]
```diff
@@ -14,24 +17,41 @@ AS
...
+                                transferIdIssuer varchar(50), transferIdMerchant varchar(50), transferDateTime datetime, localDateTime varchar(14), issuerSettlementDate date, channelId bigint,
+                                channelType varchar(50), ordererId bigint, merchantId varchar(50), merchantInvoice varchar(50), merchantPort varchar(50), merchantType varchar(50),
+                                cardId bigint, sourceAccount varchar(50), destinationAccount varchar(50), expireTime datetime, expireCount int, reversed bit, retryTime datetime,
+                                retryCount int, ledgerTxState smallint, issuerTxState smallint, acquirerTxState smallint, merchantTxState smallint, issuerId varchar(50), ledgerId varchar(50),
+                                transferCurrency varchar(3), transferAmount money,  acquirerFee money, issuerFee money, transferFee money,[description] varchar(250), issuerPort varchar(50),
+                                ledgerPort varchar(50),udfAcquirer xml)
+    
+    INSERT INTO @transfer
```
Comment (kalinkrustev) [inline api/sql/schema/750$transfer.transfer.get.sql]: There is no need of temp table variable, just variable holding the transferId:
```SQL
SELECT TOP 1 @transferId = t.transferId, t.transferTypeId, t.acquirerCode, 
FROM ...
```

Comment (georgikocev) [inline api/sql/schema/750$transfer.transfer.get.sql]: A SELECT statement that assigns a value to a variable must not be combined with data-retrieval operations.

**Pull request merged**